Error associating multiple floating IPs to VM with more than one NIC

## Description
When creating a VM with two NICs and trying to associate two floating IP addresses, one to each NIC, the operation fails and the VM is left with one NIC bound to a floating IP, the other unbound.

We can consistently reproduce the problem applying the following steps:
1. create a VM with two NICs
2. get two floating IPs from a pool
3. assign each floating IP to a different NIC

### Terraform Version
`Terraform v0.9.8` against an OpenStack "Newton" (Canonical distribution)

### Affected Resource
- openstack_compute_floatingip_associate_v2

### Terraform Configuration Files
```hcl
########################################################
# 1. Create the VM instance ***with 2 NICs***
#
resource "openstack_compute_instance_v2" "terraform-test" {
  name            = "floating_ip_test"
  image_name      = "${var.image_name}"
  flavor_name     = "${var.instance_flavor}"  
  #  
  # NOTE: this VM has two NICs!!
  #
  network  {
    uuid = "${openstack_networking_network_v2.network_1.id}"
  }
  network  {
	uuid = "${openstack_networking_network_v2.network_1.id}"
  }
}

########################################################
# 2. Retrieve two floating IPs from pool
#
resource "openstack_compute_floatingip_v2" "fips" {
  count			  = "2"
  pool = "${var.floatingIpPool}"
}

########################################################
# 3. Attempt to assign a different floating IP to each of the two NICs
#
resource "openstack_compute_floatingip_associate_v2" "fip_ass0" { 
  floating_ip = "${openstack_compute_floatingip_v2.fips.0.address}"
  instance_id = "${openstack_compute_instance_v2.terraform-test.id}"
  fixed_ip    = "${ openstack_compute_instance_v2.terraform-test.network.0.fixed_ip_v4 }"
}

# NOTE: uncomment this to reproduce the failure
/*
resource "openstack_compute_floatingip_associate_v2" "fip_ass1" {
  floating_ip = "${openstack_compute_floatingip_v2.fips.1.address}"
  instance_id = "${openstack_compute_instance_v2.terraform-test.id}"
  fixed_ip    = "${ openstack_compute_instance_v2.terraform-test.network.1.fixed_ip_v4 }"
}
*/

########################################################
# TESTBED SETUP:
# 1. create a network
# 2. create a subnet with it
# 3. create a router with external gateway
# 4. attach the network to the router

########################################################
# 1. create a network
#
resource "openstack_networking_network_v2" "network_1" {
  name           = "terraform_network"
  admin_state_up = "true"
}

########################################################
# 2. create the subnet
#
resource "openstack_networking_subnet_v2" "subnet_1" {
  name            = "terraform_subnet"
  network_id      = "${openstack_networking_network_v2.network_1.id}"
  cidr            = "192.168.0.0/24"
  ip_version      = 4
}

########################################################
# 3. create the router
#
resource "openstack_networking_router_v2" "external_router" {
  name             = "terraform_router"
  admin_state_up   = "true"
 #add provider network as external interface
  external_gateway = "${var.ext_gateway_id}"
}

########################################################
# 4. subnet_1 to router interfaces
#
resource "openstack_networking_router_interface_v2" "router_interface_to_subnet" {
  router_id = "${openstack_networking_router_v2.external_router.id}"
  subnet_id = "${openstack_networking_subnet_v2.subnet_1.id}"
}
 
########################################################
# GENERAL CONFIGURATION VARIABLES
#

variable "open_stack_end_point" {
	default = "http://X.Y.Z.W:5000/v2.0"
}
variable "tenant" {
	default = "tenant-XXXX"
}
variable "access_key" {
	default = "XXXX"
}
variable "secret_key" {
	default = "XXXX"
}

########################################################
# Our network 
#
variable "ext_gateway_id" {
	default = "XXXX"
}
variable "floatingIpPool" {
	default = "XXXX"
}

########################################################
# Types of VM  instances
#
variable "image_name" {
	default = "xenial"
}
variable "instance_flavor" {
	default = "m1.small"
}

########################################################
# Configure the OpenStack Provider
#
provider "openstack" {
  user_name   = "${var.access_key}"
  tenant_name = "${var.tenant}"
  password    = "${var.secret_key}"
  auth_url    = "${var.open_stack_end_point}"
}
```

### Debug Output
These are the relevant portions of the debug output; if tou need the complete output please advise.
```
[...]
2017/06/16 10:01:39 [DEBUG] apply: openstack_compute_floatingip_associate_v2.fip_ass1: executing Apply
2017/06/16 10:01:39 [DEBUG] plugin: terraform.exe: openstack-provider (internal) 2017/06/16 10:01:39 [DEBUG] No meta timeoutkey found in Apply()
2017/06/16 10:01:39 [DEBUG] plugin: terraform.exe: openstack-provider (internal) 2017/06/16 10:01:39 [DEBUG] Associate Options: floatingips.AssociateOpts{FloatingIP:"10.114.0.15", FixedIP:"192.168.0.9"}
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalReadDiff
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalCompareDiff
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalGetProvider
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalReadState
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalApplyPre
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalApply
2017/06/16 10:01:39 [DEBUG] apply: openstack_compute_floatingip_associate_v2.fip_ass0: executing Apply
2017/06/16 10:01:39 [DEBUG] plugin: terraform.exe: openstack-provider (internal) 2017/06/16 10:01:39 [DEBUG] No meta timeoutkey found in Apply()
2017/06/16 10:01:39 [DEBUG] plugin: terraform.exe: openstack-provider (internal) 2017/06/16 10:01:39 [DEBUG] Associate Options: floatingips.AssociateOpts{FloatingIP:"10.114.0.16", FixedIP:"192.168.0.9"}
2017/06/16 10:01:39 [DEBUG] dag/walk: vertex "root", waiting for: "provider.openstack (close)"
2017/06/16 10:01:39 [DEBUG] dag/walk: vertex "meta.count-boundary (count boundary fixup)", waiting for: "openstack_compute_floatingip_associate_v2.fip_ass0"
2017/06/16 10:01:39 [DEBUG] dag/walk: vertex "provider.openstack (close)", waiting for: "openstack_compute_floatingip_associate_v2.fip_ass1"
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalWriteState
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalApplyProvisioners
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalIf
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalWriteState
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalWriteDiff
2017/06/16 10:01:39 [DEBUG] root: eval: *terraform.EvalApplyPost
2017/06/16 10:01:39 [ERROR] root: eval: *terraform.EvalApplyPost, err: 1 error(s) occurred:

* openstack_compute_floatingip_associate_v2.fip_ass1: Error associating Floating IP: Invalid request due to incorrect syntax or missing required parameters.
2017/06/16 10:01:39 [ERROR] root: eval: *terraform.EvalSequence, err: 1 error(s) occurred:

* openstack_compute_floatingip_associate_v2.fip_ass1: Error associating Floating IP: Invalid request due to incorrect syntax or missing required parameters.
2017/06/16 10:01:39 [TRACE] [walkApply] Exiting eval tree: openstack_compute_floatingip_associate_v2.fip_ass1
2017/06/16 10:01:41 [DEBUG] plugin: terraform.exe: openstack-provider (internal) 2017/06/16 10:01:41 [DEBUG] Checking for Floating IP existence via Network API
2017/06/16 10:01:41 [DEBUG] root: eval: *terraform.EvalWriteState
2017/06/16 10:01:41 [DEBUG] root: eval: *terraform.EvalApplyProvisioners
[...]
```

### Expected Behavior
Associate each FloatingIP to a different NIC.

### Actual Behavior
Only the first association succeeds, the system has a partial state update and the following message is printed to screen:
```
Error applying plan:

1 error(s) occurred:

* openstack_compute_floatingip_associate_v2.fip_ass0: 1 error(s) occurred:

* openstack_compute_floatingip_associate_v2.fip_ass0: Error associating Floating
 IP: Invalid request due to incorrect syntax or missing required parameters.

Terraform does not automatically rollback in the face of errors.
Instead, your Terraform state file has been partially updated with
any resources that successfully completed. Please address the error
above and apply again to incrementally change your infrastructure.
```

### Steps to Reproduce
`terraform apply` with the given .tf file

### Important Factoids
The same failure occurs whether you use the "association" resource or you provide the floating IP attribute directly in the network block of the VM instance (which is deprecated, though).

--- a/openstack/compute_instance_v2_networking.go
+++ b/openstack/compute_instance_v2_networking.go
@@ -1,0 +1,419 @@
+// This set of code handles all functions required to configure networking
+// on an openstack_compute_instance_v2 resource.
+//
+// This is a complicated task because it's not possible to obtain all
+// information in a single API call. In fact, it even traverses multiple
+// OpenStack services.
+//
+// The end result, from the user's point of view, is a structured set of
+// understandable network information within the instance resource.
+package openstack
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/tenantnetworks"
+	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+// InstanceNIC is a structured representation of a Gophercloud servers.Server
+// virtual NIC.
+type InstanceNIC struct {
+	FixedIPv4 string
+	FixedIPv6 string
+	MAC       string
+}
+
+// InstanceAddresses is a collection of InstanceNICs, grouped by the
+// network name. An instance/server could have multiple NICs on the same
+// network.
+type InstanceAddresses struct {
+	NetworkName  string
+	InstanceNICs []InstanceNIC
+}
+
+// InstanceNetwork represents a collection of network information that a
+// Terraform instance needs to satisfy all network information requirements.
+type InstanceNetwork struct {
+	UUID          string
+	Name          string
+	Port          string
+	FixedIP       string
+	AccessNetwork bool
+}
+
+// getAllInstanceNetworks loops through the networks defined in the Terraform
+// configuration and structures that information into something standard that
+// can be consumed by both OpenStack and Terraform.
+//
+// This would be simple, except we have ensure both the network name and
+// network ID have been determined. This isn't just for the convenience of a
+// user specifying a human-readable network name, but the network information
+// returned by an OpenStack instance only has the network name set! So if a
+// user specified a network ID, there's no way to correlate it to the instance
+// unless we know both the name and ID.
+//
+// Not only that, but we have to account for two OpenStack network services
+// running: nova-network (legacy) and Neutron (current).
+//
+// So, let's begin the journey.
+func getAllInstanceNetworks(d *schema.ResourceData, meta interface{}) ([]InstanceNetwork, error) {
+	var instanceNetworks []InstanceNetwork
+
+	networks := d.Get("network").([]interface{})
+	for _, v := range networks {
+		network := v.(map[string]interface{})
+		networkID := network["uuid"].(string)
+		networkName := network["name"].(string)
+
+		if networkID == "" && networkName == "" {
+			return nil, fmt.Errorf("At least one of network.uuid or network.name must be set.")
+		}
+
+		// If a user specified both an ID and name, that makes things easy
+		// since both name and ID are already satisfied. No need to query
+		// further.
+		if networkID != "" && networkName != "" {
+			v := InstanceNetwork{
+				UUID:          networkID,
+				Name:          networkName,
+				Port:          network["port"].(string),
+				FixedIP:       network["fixed_ip_v4"].(string),
+				AccessNetwork: network["access_network"].(bool),
+			}
+			instanceNetworks = append(instanceNetworks, v)
+			continue
+		}
+
+		// But if at least one of name or ID was missing, we have to query
+		// for that other piece.
+		//
+		// Priority is giving to the network ID since it's guaranteed to be
+		// an exact match.
+		queryType := "name"
+		queryTerm := networkName
+		if networkID != "" {
+			queryType = "id"
+			queryTerm = networkID
+		}
+
+		networkInfo, err := getInstanceNetworkInfo(d, meta, queryType, queryTerm)
+		if err != nil {
+			return nil, err
+		}
+
+		v := InstanceNetwork{
+			UUID:          networkInfo["uuid"].(string),
+			Name:          networkInfo["name"].(string),
+			Port:          network["port"].(string),
+			FixedIP:       network["fixed_ip_v4"].(string),
+			AccessNetwork: network["access_network"].(bool),
+		}
+
+		instanceNetworks = append(instanceNetworks, v)
+	}
+
+	log.Printf("[DEBUG] getAllInstanceNetworks: %#v", instanceNetworks)
+	return instanceNetworks, nil
+}
+
+// getInstanceNetworkInfo will query for network information in order to make
+// an accurate determination of a network's name and a network's ID.
+//
+// We will try to first query the Neutron network service and fall back to the
+// legacy nova-network service if that fails.
+//
+// If OS_NOVA_NETWORK is set, query nova-network even if Neutron is available.
+// This is to be able to explicitly test the nova-network API.
+func getInstanceNetworkInfo(
+	d *schema.ResourceData, meta interface{}, queryType, queryTerm string) (map[string]interface{}, error) {
+
+	config := meta.(*Config)
+
+	if _, ok := os.LookupEnv("OS_NOVA_NETWORK"); !ok {
+		networkClient, err := config.networkingV2Client(GetRegion(d, config))
+		if err == nil {
+			networkInfo, err := getInstanceNetworkInfoNeutron(networkClient, queryType, queryTerm)
+			if err != nil {
+				return nil, fmt.Errorf("Error trying to get network information from the Network API: %s", err)
+			}
+
+			return networkInfo, nil
+		}
+	}
+
+	log.Printf("[DEBUG] Unable to obtain a network client")
+
+	computeClient, err := config.computeV2Client(GetRegion(d, config))
+	if err != nil {
+		return nil, fmt.Errorf("Error creating OpenStack compute client: %s", err)
+	}
+
+	networkInfo, err := getInstanceNetworkInfoNovaNet(computeClient, queryType, queryTerm)
+	if err != nil {
+		return nil, fmt.Errorf("Error trying to get network information from the Nova API: %s", err)
+	}
+
+	return networkInfo, nil
+}
+
+// getInstanceNetworkInfoNovaNet will query the os-tenant-networks API for
+// the network information.
+func getInstanceNetworkInfoNovaNet(
+	client *gophercloud.ServiceClient, queryType, queryTerm string) (map[string]interface{}, error) {
+
+	allPages, err := tenantnetworks.List(client).AllPages()
+	if err != nil {
+		return nil, fmt.Errorf(
+			"An error occured while querying the Nova API for network information: %s", err)
+	}
+
+	networkList, err := tenantnetworks.ExtractNetworks(allPages)
+	if err != nil {
+		return nil, fmt.Errorf(
+			"An error occured while querying the Nova API for network information: %s", err)
+	}
+
+	var networkFound bool
+	var network tenantnetworks.Network
+
+	for _, v := range networkList {
+		if queryType == "id" && v.ID == queryTerm {
+			networkFound = true
+			network = v
+			break
+		}
+
+		if queryType == "name" && v.Name == queryTerm {
+			networkFound = true
+			network = v
+			break
+		}
+	}
+
+	if networkFound {
+		v := map[string]interface{}{
+			"uuid": network.ID,
+			"name": network.Name,
+		}
+
+		log.Printf("[DEBUG] getInstanceNetworkInfoNovaNet: %#v", v)
+		return v, nil
+	}
+
+	return nil, fmt.Errorf("Could not find any matching network for %s %s", queryType, queryTerm)
+}
+
+// getInstanceNetworkInfoNeutron will query the neutron API for the network
+// information.
+func getInstanceNetworkInfoNeutron(
+	client *gophercloud.ServiceClient, queryType, queryTerm string) (map[string]interface{}, error) {
+
+	listOpts := networks.ListOpts{
+		Status: "ACTIVE",
+	}
+
+	switch queryType {
+	case "name":
+		listOpts.Name = queryTerm
+	default:
+		listOpts.ID = queryTerm
+	}
+
+	allPages, err := networks.List(client, listOpts).AllPages()
+	if err != nil {
+		return nil, fmt.Errorf("Unable to retrieve networks from the Network API: %s", err)
+	}
+
+	allNetworks, err := networks.ExtractNetworks(allPages)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to retrieve networks from the Network API: %s", err)
+	}
+
+	var network networks.Network
+	switch len(allNetworks) {
+	case 0:
+		return nil, fmt.Errorf("Could not find any matching network for %s %s", queryType, queryTerm)
+	case 1:
+		network = allNetworks[0]
+	default:
+		return nil, fmt.Errorf("More than one network found for %s %s", queryType, queryTerm)
+	}
+
+	v := map[string]interface{}{
+		"uuid": network.ID,
+		"name": network.Name,
+	}
+
+	log.Printf("[DEBUG] getInstanceNetworkInfoNeutron: %#v", v)
+	return v, nil
+}
+
+// getInstanceAddresses parses a Gophercloud server.Server's Address field into
+// a structured InstanceAddresses struct.
+func getInstanceAddresses(addresses map[string]interface{}) []InstanceAddresses {
+	var allInstanceAddresses []InstanceAddresses
+
+	for networkName, v := range addresses {
+		instanceAddresses := InstanceAddresses{
+			NetworkName: networkName,
+		}
+
+		for _, v := range v.([]interface{}) {
+			instanceNIC := InstanceNIC{}
+			v := v.(map[string]interface{})
+			if v["OS-EXT-IPS:type"] == "fixed" {
+				switch v["version"].(float64) {
+				case 6:
+					instanceNIC.FixedIPv6 = fmt.Sprintf("[%s]", v["addr"].(string))
+				default:
+					instanceNIC.FixedIPv4 = v["addr"].(string)
+				}
+			}
+
+			if v, ok := v["OS-EXT-IPS-MAC:mac_addr"].(string); ok {
+				instanceNIC.MAC = v
+			}
+
+			instanceAddresses.InstanceNICs = append(instanceAddresses.InstanceNICs, instanceNIC)
+		}
+
+		allInstanceAddresses = append(allInstanceAddresses, instanceAddresses)
+	}
+
+	log.Printf("[DEBUG] Addresses: %#v", addresses)
+	log.Printf("[DEBUG] allInstanceAddresses: %#v", allInstanceAddresses)
+
+	return allInstanceAddresses
+}
+
+// expandInstanceNetworks takes network information found in []InstanceNetwork
+// and builds a Gophercloud []servers.Network for use in creating an Instance.
+func expandInstanceNetworks(allInstanceNetworks []InstanceNetwork) []servers.Network {
+	var networks []servers.Network
+	for _, v := range allInstanceNetworks {
+		n := servers.Network{
+			UUID:    v.UUID,
+			Port:    v.Port,
+			FixedIP: v.FixedIP,
+		}
+		networks = append(networks, n)
+	}
+
+	return networks
+}
+
+// flattenInstanceNetworks collects instance network information from different
+// sources and aggregates it all together into a map array.
+func flattenInstanceNetworks(
+	d *schema.ResourceData, meta interface{}) ([]map[string]interface{}, error) {
+
+	config := meta.(*Config)
+	computeClient, err := config.computeV2Client(GetRegion(d, config))
+	if err != nil {
+		return nil, fmt.Errorf("Error creating OpenStack compute client: %s", err)
+	}
+
+	server, err := servers.Get(computeClient, d.Id()).Extract()
+	if err != nil {
+		return nil, CheckDeleted(d, err, "server")
+	}
+
+	allInstanceAddresses := getInstanceAddresses(server.Addresses)
+	allInstanceNetworks, err := getAllInstanceNetworks(d, meta)
+	if err != nil {
+		return nil, err
+	}
+
+	networks := []map[string]interface{}{}
+
+	// If there were no instance networks returned, this means that there
+	// was not a network specified in the Terraform configuration. When this
+	// happens, the instance will be launched on a "default" network, if one
+	// is available. If there isn't, the instance will fail to launch, so
+	// this is a safe assumption at this point.
+	if len(allInstanceNetworks) == 0 {
+		for _, instanceAddresses := range allInstanceAddresses {
+			for _, instanceNIC := range instanceAddresses.InstanceNICs {
+				v := map[string]interface{}{
+					"name":        instanceAddresses.NetworkName,
+					"fixed_ip_v4": instanceNIC.FixedIPv4,
+					"fixed_ip_v6": instanceNIC.FixedIPv6,
+					"mac":         instanceNIC.FixedIPv6,
+				}
+				networks = append(networks, v)
+			}
+		}
+
+		log.Printf("[DEBUG] getInstanceNetworksAndAddresses: %#v", networks)
+		return networks, nil
+	}
+
+	// Loop through all networks and addresses, merge relevant address details.
+	for _, instanceNetwork := range allInstanceNetworks {
+		for _, instanceAddresses := range allInstanceAddresses {
+			if instanceNetwork.Name == instanceAddresses.NetworkName {
+				// Only use one NIC since it's possible the user defined another NIC
+				// on this same network in another Terraform network block.
+				instanceNIC := instanceAddresses.InstanceNICs[0]
+				copy(instanceAddresses.InstanceNICs, instanceAddresses.InstanceNICs[1:])
+				v := map[string]interface{}{
+					"name":           instanceAddresses.NetworkName,
+					"fixed_ip_v4":    instanceNIC.FixedIPv4,
+					"fixed_ip_v6":    instanceNIC.FixedIPv6,
+					"mac":            instanceNIC.FixedIPv6,
+					"uuid":           instanceNetwork.UUID,
+					"port":           instanceNetwork.Port,
+					"access_network": instanceNetwork.AccessNetwork,
+				}
+				networks = append(networks, v)
+			}
+		}
+	}
+
+	log.Printf("[DEBUG] flattenInstanceNetworks: %#v", networks)
+	return networks, nil
+}
+
+// getInstanceAccessAddresses determines the best IP address to communicate
+// with the instance. It does this by looping through all networks and looking
+// for a valid IP address. Priority is given to a network that was flagged as
+// an access_network.
+func getInstanceAccessAddresses(
+	d *schema.ResourceData, networks []map[string]interface{}) (string, string) {
+
+	var hostv4, hostv6 string
+
+	// Loop through all networks
+	// If the network has a valid fixed v4 or fixed v6 address
+	// and hostv4 or hostv6 is not set, set hostv4/hostv6.
+	// If the network is an "access_network" overwrite hostv4/hostv6.
+	for _, n := range networks {
+		var accessNetwork bool
+
+		if an, ok := n["access_network"].(bool); ok && an {
+			accessNetwork = true
+		}
+
+		if fixedIPv4, ok := n["fixed_ip_v4"].(string); ok && fixedIPv4 != "" {
+			if hostv4 == "" || accessNetwork {
+				hostv4 = fixedIPv4
+			}
+		}
+
+		if fixedIPv6, ok := n["fixed_ip_v6"].(string); ok && fixedIPv6 != "" {
+			if hostv6 == "" || accessNetwork {
+				hostv6 = fixedIPv6
+			}
+		}
+	}
+
+	log.Printf("[DEBUG] OpenStack Instance Network Access Addresses: %s, %s", hostv4, hostv6)
+
+	return hostv4, hostv6
+}

--- a/openstack/compute_instance_v2_networking.go
+++ b/openstack/compute_instance_v2_networking.go
@@ -18,6 +18,7 @@ import (
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/tenantnetworks"
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
 	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
 	"github.com/hashicorp/terraform/helper/schema"
 )
 
@@ -61,6 +62,11 @@ type InstanceNetwork struct {
 // Not only that, but we have to account for two OpenStack network services
 // running: nova-network (legacy) and Neutron (current).
 //
+// In addition, if a port was specified, not all of the port information
+// will be displayed, such as multiple fixed and floating IPs. This resource
+// isn't currently configured for that type of flexibility. It's better to
+// reference the actual port resource itself.
+//
 // So, let's begin the journey.
 func getAllInstanceNetworks(d *schema.ResourceData, meta interface{}) ([]InstanceNetwork, error) {
 	var instanceNetworks []InstanceNetwork
@@ -70,9 +76,11 @@ func getAllInstanceNetworks(d *schema.ResourceData, meta interface{}) ([]Instanc
 		network := v.(map[string]interface{})
 		networkID := network["uuid"].(string)
 		networkName := network["name"].(string)
+		portID := network["port"].(string)
 
-		if networkID == "" && networkName == "" {
-			return nil, fmt.Errorf("At least one of network.uuid or network.name must be set.")
+		if networkID == "" && networkName == "" && portID == "" {
+			return nil, fmt.Errorf(
+				"At least one of network.uuid, network.name, or network.port must be set.")
 		}
 
 		// If a user specified both an ID and name, that makes things easy
@@ -82,7 +90,7 @@ func getAllInstanceNetworks(d *schema.ResourceData, meta interface{}) ([]Instanc
 			v := InstanceNetwork{
 				UUID:          networkID,
 				Name:          networkName,
-				Port:          network["port"].(string),
+				Port:          portID,
 				FixedIP:       network["fixed_ip_v4"].(string),
 				AccessNetwork: network["access_network"].(bool),
 			}
@@ -93,13 +101,20 @@ func getAllInstanceNetworks(d *schema.ResourceData, meta interface{}) ([]Instanc
 		// But if at least one of name or ID was missing, we have to query
 		// for that other piece.
 		//
-		// Priority is giving to the network ID since it's guaranteed to be
+		// Priority is given to a port since a network ID or name usually isn't
+		// specified when using a port.
+		//
+		// Next priority is given to the network ID since it's guaranteed to be
 		// an exact match.
 		queryType := "name"
 		queryTerm := networkName
 		if networkID != "" {
 			queryType = "id"
 			queryTerm = networkID
+		}
+		if portID != "" {
+			queryType = "port"
+			queryTerm = portID
 		}
 
 		networkInfo, err := getInstanceNetworkInfo(d, meta, queryType, queryTerm)
@@ -110,7 +125,7 @@ func getAllInstanceNetworks(d *schema.ResourceData, meta interface{}) ([]Instanc
 		v := InstanceNetwork{
 			UUID:          networkInfo["uuid"].(string),
 			Name:          networkInfo["name"].(string),
-			Port:          network["port"].(string),
+			Port:          portID,
 			FixedIP:       network["fixed_ip_v4"].(string),
 			AccessNetwork: network["access_network"].(bool),
 		}
@@ -167,6 +182,12 @@ func getInstanceNetworkInfo(
 func getInstanceNetworkInfoNovaNet(
 	client *gophercloud.ServiceClient, queryType, queryTerm string) (map[string]interface{}, error) {
 
+	// If somehow a port ended up here, we should just error out.
+	if queryType == "port" {
+		return nil, fmt.Errorf(
+			"Unable to query a port (%s) using the Nova API", queryTerm)
+	}
+
 	allPages, err := tenantnetworks.List(client).AllPages()
 	if err != nil {
 		return nil, fmt.Errorf(
@@ -213,6 +234,36 @@ func getInstanceNetworkInfoNovaNet(
 // information.
 func getInstanceNetworkInfoNeutron(
 	client *gophercloud.ServiceClient, queryType, queryTerm string) (map[string]interface{}, error) {
+
+	// If a port was specified, use it to look up the network ID
+	// and then query the network as if a network ID was originally used.
+	if queryType == "port" {
+		listOpts := ports.ListOpts{
+			ID: queryTerm,
+		}
+		allPages, err := ports.List(client, listOpts).AllPages()
+		if err != nil {
+			return nil, fmt.Errorf("Unable to retrieve networks from the Network API: %s", err)
+		}
+
+		allPorts, err := ports.ExtractPorts(allPages)
+		if err != nil {
+			return nil, fmt.Errorf("Unable to retrieve networks from the Network API: %s", err)
+		}
+
+		var port ports.Port
+		switch len(allPorts) {
+		case 0:
+			return nil, fmt.Errorf("Could not find any matching port for %s %s", queryType, queryTerm)
+		case 1:
+			port = allPorts[0]
+		default:
+			return nil, fmt.Errorf("More than one port found for %s %s", queryType, queryTerm)
+		}
+
+		queryType = "id"
+		queryTerm = port.NetworkID
+	}
 
 	listOpts := networks.ListOpts{
 		Status: "ACTIVE",
@@ -350,7 +401,7 @@ func flattenInstanceNetworks(
 			}
 		}
 
-		log.Printf("[DEBUG] getInstanceNetworksAndAddresses: %#v", networks)
+		log.Printf("[DEBUG] flattenInstanceNetworks: %#v", networks)
 		return networks, nil
 	}
 

--- a/openstack/resource_openstack_compute_instance_v2.go
+++ b/openstack/resource_openstack_compute_instance_v2.go
@@ -16,7 +16,6 @@ import (
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/schedulerhints"
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/secgroups"
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/startstop"
-	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/tenantnetworks"
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/flavors"
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/images"
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
@@ -375,19 +374,13 @@ func resourceComputeInstanceV2Create(d *schema.ResourceData, meta interface{}) e
 
 	// Build a list of networks with the information given upon creation.
 	// Error out if an invalid network configuration was used.
-	networkDetails, err := getInstanceNetworks(computeClient, d)
+	allInstanceNetworks, err := getAllInstanceNetworks(d, meta)
 	if err != nil {
 		return err
 	}
 
-	networks := make([]servers.Network, len(networkDetails))
-	for i, net := range networkDetails {
-		networks[i] = servers.Network{
-			UUID:    net["uuid"].(string),
-			Port:    net["port"].(string),
-			FixedIP: net["fixed_ip_v4"].(string),
-		}
-	}
+	// Build a []servers.Network to pass into the create options.
+	networks := expandInstanceNetworks(allInstanceNetworks)
 
 	configDrive := d.Get("config_drive").(bool)
 
@@ -475,10 +468,6 @@ func resourceComputeInstanceV2Create(d *schema.ResourceData, meta interface{}) e
 			server.ID, err)
 	}
 
-	// Now that the instance has been created, we need to do an early read on the
-	// networks in order to associate floating IPs
-	_, err = getInstanceNetworksAndAddresses(computeClient, d)
-
 	return resourceComputeInstanceV2Read(d, meta)
 }
 
@@ -499,7 +488,7 @@ func resourceComputeInstanceV2Read(d *schema.ResourceData, meta interface{}) err
 	d.Set("name", server.Name)
 
 	// Get the instance network and address information
-	networks, err := getInstanceNetworksAndAddresses(computeClient, d)
+	networks, err := flattenInstanceNetworks(d, meta)
 	if err != nil {
 		return err
 	}
@@ -507,6 +496,8 @@ func resourceComputeInstanceV2Read(d *schema.ResourceData, meta interface{}) err
 	// Determine the best IPv4 and IPv6 addresses to access the instance with
 	hostv4, hostv6 := getInstanceAccessAddresses(d, networks)
 
+	// AccessIPv4/v6 isn't standard in OpenStack, but there have been reports
+	// of them being used in some environments.
 	if server.AccessIPv4 != "" && hostv4 == "" {
 		hostv4 = server.AccessIPv4
 	}
@@ -521,7 +512,7 @@ func resourceComputeInstanceV2Read(d *schema.ResourceData, meta interface{}) err
 
 	// Determine the best IP address to use for SSH connectivity.
 	// Prefer IPv4 over IPv6.
-	preferredSSHAddress := ""
+	var preferredSSHAddress string
 	if hostv4 != "" {
 		preferredSSHAddress = hostv4
 	} else if hostv6 != "" {
@@ -833,202 +824,6 @@ func resourceInstanceSecGroupsV2(d *schema.ResourceData) []string {
 		secgroups[i] = raw.(string)
 	}
 	return secgroups
-}
-
-// getInstanceNetworks collects instance network information from different sources
-// and aggregates it all together.
-func getInstanceNetworksAndAddresses(computeClient *gophercloud.ServiceClient, d *schema.ResourceData) ([]map[string]interface{}, error) {
-	server, err := servers.Get(computeClient, d.Id()).Extract()
-
-	if err != nil {
-		return nil, CheckDeleted(d, err, "server")
-	}
-
-	networkDetails, err := getInstanceNetworks(computeClient, d)
-	addresses := getInstanceAddresses(server.Addresses)
-	if err != nil {
-		return nil, err
-	}
-
-	// if there are no networkDetails, make networks at least a length of 1
-	networkLength := 1
-	if len(networkDetails) > 0 {
-		networkLength = len(networkDetails)
-	}
-	networks := make([]map[string]interface{}, networkLength)
-
-	// Loop through all networks and addresses,
-	// merge relevant address details.
-	if len(networkDetails) == 0 {
-		for netName, n := range addresses {
-			networks[0] = map[string]interface{}{
-				"name":        netName,
-				"fixed_ip_v4": n["fixed_ip_v4"],
-				"fixed_ip_v6": n["fixed_ip_v6"],
-				"mac":         n["mac"],
-			}
-		}
-	} else {
-		for i, net := range networkDetails {
-			n := addresses[net["name"].(string)]
-
-			networks[i] = map[string]interface{}{
-				"uuid":           networkDetails[i]["uuid"],
-				"name":           networkDetails[i]["name"],
-				"port":           networkDetails[i]["port"],
-				"fixed_ip_v4":    n["fixed_ip_v4"],
-				"fixed_ip_v6":    n["fixed_ip_v6"],
-				"mac":            n["mac"],
-				"access_network": networkDetails[i]["access_network"],
-			}
-		}
-	}
-
-	log.Printf("[DEBUG] networks: %+v", networks)
-
-	return networks, nil
-}
-
-func getInstanceNetworks(computeClient *gophercloud.ServiceClient, d *schema.ResourceData) ([]map[string]interface{}, error) {
-	rawNetworks := d.Get("network").([]interface{})
-	newNetworks := make([]map[string]interface{}, 0, len(rawNetworks))
-	var tenantnet tenantnetworks.Network
-
-	tenantNetworkExt := true
-	for _, raw := range rawNetworks {
-		// Not sure what causes this, but it is a possibility (see GH-2323).
-		// Since we call this function to reconcile what we'll save in the
-		// state anyways, we just ignore it.
-		if raw == nil {
-			continue
-		}
-
-		rawMap := raw.(map[string]interface{})
-
-		allPages, err := tenantnetworks.List(computeClient).AllPages()
-		if err != nil {
-			if _, ok := err.(gophercloud.ErrDefault404); ok {
-				log.Printf("[DEBUG] os-tenant-networks disabled")
-				tenantNetworkExt = false
-			}
-
-			log.Printf("[DEBUG] Err looks like: %+v", err)
-			if errCode, ok := err.(gophercloud.ErrUnexpectedResponseCode); ok {
-				if errCode.Actual == 403 {
-					log.Printf("[DEBUG] os-tenant-networks disabled.")
-					tenantNetworkExt = false
-				} else {
-					log.Printf("[DEBUG] unexpected os-tenant-networks error: %s", err)
-					tenantNetworkExt = false
-				}
-			}
-		}
-
-		// In some cases, a call to os-tenant-networks might work,
-		// but the response is invalid. Catch this during extraction.
-		networkList := []tenantnetworks.Network{}
-		if tenantNetworkExt {
-			networkList, err = tenantnetworks.ExtractNetworks(allPages)
-			if err != nil {
-				log.Printf("[DEBUG] error extracting os-tenant-networks results: %s", err)
-				tenantNetworkExt = false
-			}
-		}
-
-		networkID := ""
-		networkName := ""
-		networkFound := false
-		if tenantNetworkExt {
-			for _, network := range networkList {
-				if network.Name == rawMap["name"] {
-					tenantnet = network
-					networkFound = true
-				}
-				if network.ID == rawMap["uuid"] {
-					tenantnet = network
-					networkFound = true
-				}
-			}
-
-			networkID = tenantnet.ID
-			networkName = tenantnet.Name
-		} else {
-			networkID = rawMap["uuid"].(string)
-			networkName = rawMap["name"].(string)
-			networkFound = true
-		}
-
-		if networkFound {
-			newNetworks = append(newNetworks, map[string]interface{}{
-				"uuid":           networkID,
-				"name":           networkName,
-				"port":           rawMap["port"].(string),
-				"fixed_ip_v4":    rawMap["fixed_ip_v4"].(string),
-				"access_network": rawMap["access_network"].(bool),
-			})
-		} else {
-			log.Printf("[DEBUG] Unable to find a network with the name [%s] or the uuid [%s]", rawMap["name"], rawMap["uuid"])
-		}
-	}
-
-	log.Printf("[DEBUG] networks: %+v", newNetworks)
-	return newNetworks, nil
-}
-
-func getInstanceAddresses(addresses map[string]interface{}) map[string]map[string]interface{} {
-	addrs := make(map[string]map[string]interface{})
-	for n, networkAddresses := range addresses {
-		addrs[n] = make(map[string]interface{})
-		for _, element := range networkAddresses.([]interface{}) {
-			address := element.(map[string]interface{})
-			if address["OS-EXT-IPS:type"] == "fixed" {
-				if address["version"].(float64) == 4 {
-					addrs[n]["fixed_ip_v4"] = address["addr"].(string)
-				} else {
-					addrs[n]["fixed_ip_v6"] = fmt.Sprintf("[%s]", address["addr"].(string))
-				}
-			}
-			if mac, ok := address["OS-EXT-IPS-MAC:mac_addr"]; ok {
-				addrs[n]["mac"] = mac.(string)
-			}
-		}
-	}
-
-	log.Printf("[DEBUG] Addresses: %+v", addresses)
-
-	return addrs
-}
-
-func getInstanceAccessAddresses(d *schema.ResourceData, networks []map[string]interface{}) (string, string) {
-	var hostv4, hostv6 string
-
-	// Loop through all networks
-	// If the network has a valid fixed v4 or fixed v6 address
-	// and hostv4 or hostv6 is not set, set hostv4/hostv6.
-	// If the network is an "access_network" overwrite hostv4/hostv6.
-	for _, n := range networks {
-		var accessNetwork bool
-
-		if an, ok := n["access_network"].(bool); ok && an {
-			accessNetwork = true
-		}
-
-		if fixedIPv4, ok := n["fixed_ip_v4"].(string); ok && fixedIPv4 != "" {
-			if hostv4 == "" || accessNetwork {
-				hostv4 = fixedIPv4
-			}
-		}
-
-		if fixedIPv6, ok := n["fixed_ip_v6"].(string); ok && fixedIPv6 != "" {
-			if hostv6 == "" || accessNetwork {
-				hostv6 = fixedIPv6
-			}
-		}
-	}
-
-	log.Printf("[DEBUG] OpenStack Instance Network Access Addresses: %s, %s", hostv4, hostv6)
-
-	return hostv4, hostv6
 }
 
 func resourceInstanceMetadataV2(d *schema.ResourceData) map[string]string {

--- a/openstack/resource_openstack_compute_instance_v2_test.go
+++ b/openstack/resource_openstack_compute_instance_v2_test.go
@@ -13,6 +13,7 @@ import (
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/extensions/volumeattach"
 	"github.com/gophercloud/gophercloud/openstack/compute/v2/servers"
 	"github.com/gophercloud/gophercloud/openstack/networking/v2/networks"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/ports"
 	"github.com/gophercloud/gophercloud/pagination"
 )
 
@@ -438,6 +439,32 @@ func TestAccComputeV2Instance_crazyNICs(t *testing.T) {
 						"openstack_compute_instance_v2.instance_1", "network.3.fixed_ip_v4", "192.168.1.101"),
 					resource.TestCheckResourceAttr(
 						"openstack_compute_instance_v2.instance_1", "network.4.fixed_ip_v4", "192.168.2.101"),
+				),
+			},
+		},
+	})
+}
+
+func TestAccComputeV2Instance_networkPort(t *testing.T) {
+	var instance servers.Server
+	var network networks.Network
+	var port ports.Port
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckComputeV2InstanceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccComputeV2Instance_networkPort,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckComputeV2InstanceExists("openstack_compute_instance_v2.instance_1", &instance),
+					testAccCheckNetworkingV2NetworkExists(
+						"openstack_networking_network_v2.network_1", &network),
+					testAccCheckNetworkingV2PortExists(
+						"openstack_networking_port_v2.port_1", &port),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.0.port", &port.ID),
 				),
 			},
 		},
@@ -1037,3 +1064,42 @@ resource "openstack_compute_instance_v2" "instance_1" {
   }
 }
 `, OS_NETWORK_ID)
+
+const testAccComputeV2Instance_networkPort = `
+resource "openstack_networking_network_v2" "network_1" {
+  name = "network_1"
+}
+
+resource "openstack_networking_subnet_v2" "subnet_1" {
+  name = "subnet_1"
+  network_id = "${openstack_networking_network_v2.network_1.id}"
+  cidr = "192.168.1.0/24"
+  ip_version = 4
+  enable_dhcp = true
+  no_gateway = true
+}
+
+resource "openstack_networking_port_v2" "port_1" {
+  name = "port_1"
+  network_id = "${openstack_networking_network_v2.network_1.id}"
+  admin_state_up = "true"
+
+  fixed_ip {
+    subnet_id = "${openstack_networking_subnet_v2.subnet_1.id}"
+    ip_address = "192.168.1.100"
+  }
+}
+
+resource "openstack_compute_instance_v2" "instance_1" {
+  depends_on = [
+    "openstack_networking_port_v2.port_1"
+  ]
+
+  name = "instance_1"
+  security_groups = ["default"]
+
+  network {
+    port = "${openstack_networking_port_v2.port_1.id}"
+  }
+}
+`

--- a/openstack/resource_openstack_compute_instance_v2_test.go
+++ b/openstack/resource_openstack_compute_instance_v2_test.go
@@ -401,6 +401,10 @@ func TestAccComputeV2Instance_crazyNICs(t *testing.T) {
 	var instance servers.Server
 	var network_1 networks.Network
 	var network_2 networks.Network
+	var port_1 ports.Port
+	var port_2 ports.Port
+	var port_3 ports.Port
+	var port_4 ports.Port
 
 	resource.Test(t, resource.TestCase{
 		PreCheck:     func() { testAccPreCheck(t) },
@@ -415,6 +419,14 @@ func TestAccComputeV2Instance_crazyNICs(t *testing.T) {
 						"openstack_networking_network_v2.network_1", &network_1),
 					testAccCheckNetworkingV2NetworkExists(
 						"openstack_networking_network_v2.network_2", &network_2),
+					testAccCheckNetworkingV2PortExists(
+						"openstack_networking_port_v2.port_1", &port_1),
+					testAccCheckNetworkingV2PortExists(
+						"openstack_networking_port_v2.port_2", &port_2),
+					testAccCheckNetworkingV2PortExists(
+						"openstack_networking_port_v2.port_3", &port_3),
+					testAccCheckNetworkingV2PortExists(
+						"openstack_networking_port_v2.port_4", &port_4),
 					resource.TestCheckResourceAttrPtr(
 						"openstack_compute_instance_v2.instance_1", "network.1.uuid", &network_1.ID),
 					resource.TestCheckResourceAttrPtr(
@@ -423,6 +435,10 @@ func TestAccComputeV2Instance_crazyNICs(t *testing.T) {
 						"openstack_compute_instance_v2.instance_1", "network.3.uuid", &network_1.ID),
 					resource.TestCheckResourceAttrPtr(
 						"openstack_compute_instance_v2.instance_1", "network.4.uuid", &network_2.ID),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.5.uuid", &network_1.ID),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.6.uuid", &network_2.ID),
 					resource.TestCheckResourceAttr(
 						"openstack_compute_instance_v2.instance_1", "network.1.name", "network_1"),
 					resource.TestCheckResourceAttr(
@@ -432,6 +448,14 @@ func TestAccComputeV2Instance_crazyNICs(t *testing.T) {
 					resource.TestCheckResourceAttr(
 						"openstack_compute_instance_v2.instance_1", "network.4.name", "network_2"),
 					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.5.name", "network_1"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.6.name", "network_2"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.7.name", "network_1"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.8.name", "network_2"),
+					resource.TestCheckResourceAttr(
 						"openstack_compute_instance_v2.instance_1", "network.1.fixed_ip_v4", "192.168.1.100"),
 					resource.TestCheckResourceAttr(
 						"openstack_compute_instance_v2.instance_1", "network.2.fixed_ip_v4", "192.168.2.100"),
@@ -439,32 +463,14 @@ func TestAccComputeV2Instance_crazyNICs(t *testing.T) {
 						"openstack_compute_instance_v2.instance_1", "network.3.fixed_ip_v4", "192.168.1.101"),
 					resource.TestCheckResourceAttr(
 						"openstack_compute_instance_v2.instance_1", "network.4.fixed_ip_v4", "192.168.2.101"),
-				),
-			},
-		},
-	})
-}
-
-func TestAccComputeV2Instance_networkPort(t *testing.T) {
-	var instance servers.Server
-	var network networks.Network
-	var port ports.Port
-
-	resource.Test(t, resource.TestCase{
-		PreCheck:     func() { testAccPreCheck(t) },
-		Providers:    testAccProviders,
-		CheckDestroy: testAccCheckComputeV2InstanceDestroy,
-		Steps: []resource.TestStep{
-			resource.TestStep{
-				Config: testAccComputeV2Instance_networkPort,
-				Check: resource.ComposeTestCheckFunc(
-					testAccCheckComputeV2InstanceExists("openstack_compute_instance_v2.instance_1", &instance),
-					testAccCheckNetworkingV2NetworkExists(
-						"openstack_networking_network_v2.network_1", &network),
-					testAccCheckNetworkingV2PortExists(
-						"openstack_networking_port_v2.port_1", &port),
 					resource.TestCheckResourceAttrPtr(
-						"openstack_compute_instance_v2.instance_1", "network.0.port", &port.ID),
+						"openstack_compute_instance_v2.instance_1", "network.5.port", &port_1.ID),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.6.port", &port_2.ID),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.7.port", &port_3.ID),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.8.port", &port_4.ID),
 				),
 			},
 		},
@@ -1030,10 +1036,56 @@ resource "openstack_networking_subnet_v2" "subnet_2" {
   no_gateway = true
 }
 
+resource "openstack_networking_port_v2" "port_1" {
+  name = "port_1"
+  network_id = "${openstack_networking_network_v2.network_1.id}"
+  admin_state_up = "true"
+
+  fixed_ip {
+    subnet_id = "${openstack_networking_subnet_v2.subnet_1.id}"
+    ip_address = "192.168.1.103"
+  }
+}
+
+resource "openstack_networking_port_v2" "port_2" {
+  name = "port_2"
+  network_id = "${openstack_networking_network_v2.network_2.id}"
+  admin_state_up = "true"
+
+  fixed_ip {
+    subnet_id = "${openstack_networking_subnet_v2.subnet_2.id}"
+    ip_address = "192.168.2.103"
+  }
+}
+
+resource "openstack_networking_port_v2" "port_3" {
+  name = "port_3"
+  network_id = "${openstack_networking_network_v2.network_1.id}"
+  admin_state_up = "true"
+
+  fixed_ip {
+    subnet_id = "${openstack_networking_subnet_v2.subnet_1.id}"
+    ip_address = "192.168.1.104"
+  }
+}
+
+resource "openstack_networking_port_v2" "port_4" {
+  name = "port_4"
+  network_id = "${openstack_networking_network_v2.network_2.id}"
+  admin_state_up = "true"
+
+  fixed_ip {
+    subnet_id = "${openstack_networking_subnet_v2.subnet_2.id}"
+    ip_address = "192.168.2.104"
+  }
+}
+
 resource "openstack_compute_instance_v2" "instance_1" {
   depends_on = [
     "openstack_networking_subnet_v2.subnet_1",
     "openstack_networking_subnet_v2.subnet_2",
+    "openstack_networking_port_v2.port_1",
+    "openstack_networking_port_v2.port_2",
   ]
 
   name = "instance_1"
@@ -1062,44 +1114,21 @@ resource "openstack_compute_instance_v2" "instance_1" {
     uuid = "${openstack_networking_network_v2.network_2.id}"
     fixed_ip_v4 = "192.168.2.101"
   }
-}
-`, OS_NETWORK_ID)
-
-const testAccComputeV2Instance_networkPort = `
-resource "openstack_networking_network_v2" "network_1" {
-  name = "network_1"
-}
-
-resource "openstack_networking_subnet_v2" "subnet_1" {
-  name = "subnet_1"
-  network_id = "${openstack_networking_network_v2.network_1.id}"
-  cidr = "192.168.1.0/24"
-  ip_version = 4
-  enable_dhcp = true
-  no_gateway = true
-}
-
-resource "openstack_networking_port_v2" "port_1" {
-  name = "port_1"
-  network_id = "${openstack_networking_network_v2.network_1.id}"
-  admin_state_up = "true"
-
-  fixed_ip {
-    subnet_id = "${openstack_networking_subnet_v2.subnet_1.id}"
-    ip_address = "192.168.1.100"
-  }
-}
-
-resource "openstack_compute_instance_v2" "instance_1" {
-  depends_on = [
-    "openstack_networking_port_v2.port_1"
-  ]
-
-  name = "instance_1"
-  security_groups = ["default"]
 
   network {
     port = "${openstack_networking_port_v2.port_1.id}"
   }
+
+  network {
+    port = "${openstack_networking_port_v2.port_2.id}"
+  }
+
+  network {
+    port = "${openstack_networking_port_v2.port_3.id}"
+  }
+
+  network {
+    port = "${openstack_networking_port_v2.port_4.id}"
+  }
 }
-`
+`, OS_NETWORK_ID)

--- a/openstack/resource_openstack_compute_instance_v2_test.go
+++ b/openstack/resource_openstack_compute_instance_v2_test.go
@@ -396,6 +396,54 @@ func TestAccComputeV2Instance_networkNameToID(t *testing.T) {
 	})
 }
 
+func TestAccComputeV2Instance_crazyNICs(t *testing.T) {
+	var instance servers.Server
+	var network_1 networks.Network
+	var network_2 networks.Network
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:     func() { testAccPreCheck(t) },
+		Providers:    testAccProviders,
+		CheckDestroy: testAccCheckComputeV2InstanceDestroy,
+		Steps: []resource.TestStep{
+			resource.TestStep{
+				Config: testAccComputeV2Instance_crazyNICs,
+				Check: resource.ComposeTestCheckFunc(
+					testAccCheckComputeV2InstanceExists("openstack_compute_instance_v2.instance_1", &instance),
+					testAccCheckNetworkingV2NetworkExists(
+						"openstack_networking_network_v2.network_1", &network_1),
+					testAccCheckNetworkingV2NetworkExists(
+						"openstack_networking_network_v2.network_2", &network_2),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.1.uuid", &network_1.ID),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.2.uuid", &network_2.ID),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.3.uuid", &network_1.ID),
+					resource.TestCheckResourceAttrPtr(
+						"openstack_compute_instance_v2.instance_1", "network.4.uuid", &network_2.ID),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.1.name", "network_1"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.2.name", "network_2"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.3.name", "network_1"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.4.name", "network_2"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.1.fixed_ip_v4", "192.168.1.100"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.2.fixed_ip_v4", "192.168.2.100"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.3.fixed_ip_v4", "192.168.1.101"),
+					resource.TestCheckResourceAttr(
+						"openstack_compute_instance_v2.instance_1", "network.4.fixed_ip_v4", "192.168.2.101"),
+				),
+			},
+		},
+	})
+}
+
 func testAccCheckComputeV2InstanceDestroy(s *terraform.State) error {
 	config := testAccProvider.Meta().(*Config)
 	computeClient, err := config.computeV2Client(OS_REGION_NAME)
@@ -925,5 +973,67 @@ resource "openstack_compute_instance_v2" "instance_1" {
     name = "${openstack_networking_network_v2.network_1.name}"
   }
 
+}
+`, OS_NETWORK_ID)
+
+var testAccComputeV2Instance_crazyNICs = fmt.Sprintf(`
+resource "openstack_networking_network_v2" "network_1" {
+  name = "network_1"
+}
+
+resource "openstack_networking_subnet_v2" "subnet_1" {
+  name = "subnet_1"
+  network_id = "${openstack_networking_network_v2.network_1.id}"
+  cidr = "192.168.1.0/24"
+  ip_version = 4
+  enable_dhcp = true
+  no_gateway = true
+}
+
+resource "openstack_networking_network_v2" "network_2" {
+  name = "network_2"
+}
+
+resource "openstack_networking_subnet_v2" "subnet_2" {
+  name = "subnet_2"
+  network_id = "${openstack_networking_network_v2.network_2.id}"
+  cidr = "192.168.2.0/24"
+  ip_version = 4
+  enable_dhcp = true
+  no_gateway = true
+}
+
+resource "openstack_compute_instance_v2" "instance_1" {
+  depends_on = [
+    "openstack_networking_subnet_v2.subnet_1",
+    "openstack_networking_subnet_v2.subnet_2",
+  ]
+
+  name = "instance_1"
+  security_groups = ["default"]
+
+  network {
+    uuid = "%s"
+  }
+
+  network {
+    uuid = "${openstack_networking_network_v2.network_1.id}"
+    fixed_ip_v4 = "192.168.1.100"
+  }
+
+  network {
+    uuid = "${openstack_networking_network_v2.network_2.id}"
+    fixed_ip_v4 = "192.168.2.100"
+  }
+
+  network {
+    uuid = "${openstack_networking_network_v2.network_1.id}"
+    fixed_ip_v4 = "192.168.1.101"
+  }
+
+  network {
+    uuid = "${openstack_networking_network_v2.network_2.id}"
+    fixed_ip_v4 = "192.168.2.101"
+  }
 }
 `, OS_NETWORK_ID)

